### PR TITLE
Force index file creation on mkdocs

### PR DESCRIPTION
**Pull request by stsewd**
RTD no longer converts the README to a index file https://github.com/rtfd/readthedocs.org/commit/d907524ddec69c2530324d9c6cc7011830a209d8, when a mkdocs project is created without one, is hard to tell why the docs aren't shown (see #1615 and #3321).

### CODE REVIEW

**Comment by ericholscher** [comment]
Looks like a good PR. This is something we should figure out how to test well.

File: readthedocs/doc_builder/base.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 import logging
 import os
 import shutil
-from builtins import object
+from builtins import object, open
```
Comment (stsewd) [inline readthedocs/doc_builder/base.py]: Have to import `open` to mocking it.

File: readthedocs/rtd_tests/tests/test_builds.py [hunk 2]
```diff
@@ -54,6 +57,165 @@ def test_build(self):
...
+        version = project.versions.all()[0]
+        build_env = LocalBuildEnvironment(
+            project=project,
+            version=version,
+            build={}
+        )
+        python_env = Virtualenv(version=version, build_env=build_env)
+        base_builder = MkdocsHTML(build_env, python_env)
+
+        def look_index_path(path):
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_builds.py]: I know there is a function to fake paths based on a regex, but with this form looks more explicit, at least for me.

File: readthedocs/doc_builder/base.py (hunk 1 above)
Comment (humitos) [reply readthedocs/doc_builder/base.py]: I think this is the correct way.

File: readthedocs/rtd_tests/tests/test_builds.py [hunk 3]
```diff
@@ -54,6 +57,165 @@ def test_build(self):
...
+        )
+        python_env = Virtualenv(version=version, build_env=build_env)
+        base_builder = SphinxHTML(build_env, python_env)
+
+        def look_index_path(path):
+            if path.endswith('README.md'):
+                return True
+            elif path.endswith('index.md'):
+                return False
+            return False
```
Comment (ericholscher) [inline readthedocs/rtd_tests/tests/test_builds.py]: This is repeated 4 times, I think it should be a util method at the top of the file?

File: readthedocs/doc_builder/base.py [hunk 4]
```diff
@@ -86,37 +86,43 @@ def docs_dir(self, docs_dir=None, **__):
...
+        index_text = """
 
 Welcome to Read the Docs
 ------------------------
 
 This is an autogenerated index file.
 
-Please create an ``index.{ext}`` or ``README.{ext}`` file with your own content
-under the root (or ``/docs``) directory in your repository.
+Please create an ``index.{ext}`` or ``README.{ext}`` (Sphinx only) file with
```
Comment (ericholscher) [inline readthedocs/doc_builder/base.py]: Why is this logic Sphinx only? Do we actually do it for Sphinx projects, but not mkdocs? Should we do it for Sphinx projects?

File: readthedocs/rtd_tests/tests/test_builds.py (hunk 3 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_builds.py]: I will add this method to the mock path utils better :)

File: readthedocs/doc_builder/base.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/doc_builder/base.py]: I was thinking the same, I'm going to check if this works on Sphinx, if so, probably this `force_index` would be the default behavior.

**Comment by stsewd** [comment]
@ericholscher I take a look at the code, and notice that the template wasn't using the `master_doc` variable https://github.com/rtfd/readthedocs.org/blob/27d9ba3e9ab2a24f29f4b4cfead2ff32bebcaee6/readthedocs/templates/sphinx/conf.py.conf#L15

I did a test using the `master_doc` variable with a `README.rst` file and sphinx still gives 404. So should be `force_index` True by default? or maybe making the function always look for an `index` file?

Also notice that now with just a `README.rst` sphinx make the build to fail, since is always looking for an `index.rst` file.

**Comment by humitos** [comment]
> Also notice that now with just a README.rst sphinx make the build to fail, since is always looking for an index.rst file.

I just noticed the same in locally and I wanted to report this but it seems that you already knew it. My manual test was a simple repository with just `README.rst` file.

**Comment by stsewd** [comment]
I labeling this as `Needed: design decision` since we need to take a decision here https://github.com/rtfd/readthedocs.org/pull/3635#issuecomment-369706453 if always forcing the index file creation or use the `master_doc` variable.

File: readthedocs/doc_builder/base.py [hunk 5]
```diff
@@ -94,37 +100,46 @@ def docs_dir(self, docs_dir=None, **__):
...
+
+            Please create an ``index.{ext}`` or ``README.{ext}`` (Sphinx only) file with
+            your own content under the root (or ``/docs``) directory in your repository.
+
+            If you want to use another markup, choose a different builder in your settings.
+            Check out our `Getting Started Guide
+            <https://docs.readthedocs.io/en/latest/getting_started.html>`_ to become more
+            familiar with Read the Docs.
+            """
+        )
```
Comment (stsewd) [inline readthedocs/doc_builder/base.py]: I also realize that this will look weird on mkdocs (because of the rst specific format used here)

**Comment by stsewd** [comment]
I just discovered that Mkdocs already handle the case where a readme file is found, so we can't use the `force_index` option for mkdocs, but this PR should still be required for sphinx projects, here we want to use the `force_index` option.

**Comment by humitos** [comment]
I think we want to disable all the magic around these cases and force the user to have the same experience locally than in Read the Docs. So, if they don't have a valid index we shouldn't create one. This PR can be closed IMO.

**Comment by stsewd** [comment]
>  So, if they don't have a valid index we shouldn't create one. 

The thing is that users will get a 404 without any other information. In the generated index file we are showing a message that if they are seeing that page, they should create an index file.

**Comment by ericholscher** [comment]
Showing a different 404 page I think makes sense. We should just fix our 404 to link to the FAQ for a couple common cases. That's a different PR though.

**Comment by stsewd** [comment]
I think we should improve our 404 default page instead :+1:

**Pull request closed without merging**